src/azul/service/elasticsearch_service.py

```diff
@@ -504,61 +503,72 @@ def transform_summary(self,
+                'fileCount': 'contents.files.uuid'
+            },
+            'samples': {
+                'specimenCount': 'contents.specimens.document_id',
+                'speciesCount': 'contents.donors.genus_species',
+                'donorCount': 'contents.donors.document_id',
+            },
+            'projects': {
+                'labCount': 'contents.projects.laboratory',
+            }
+        }.get(entity_type, {})
+
+        if entity_type == 'files':
+            # Add a total file size aggregate
+            es_search.aggs.metric('total_size',
+                                  'sum',
+                                  field='contents.files.size_')
+        elif entity_type == 'cell_suspensions':
+            # Add a cell count aggregate per organ
+            es_search.aggs.bucket(
+                'group_by_organ',
+                'terms',
+                field='contents.cell_suspensions.organ.keyword',
+                size=config.terms_aggregation_size
+            ).bucket(
+                'cell_count',
+                'sum',
+                field='contents.cell_suspensions.total_estimated_cells_'
+            )
+            # Add a total cell count aggregate
+            es_search.aggs.metric('total_cell_count',
+                                  'sum',
+                                  field='contents.cell_suspensions.total_estimated_cells_')
+        elif entity_type == 'samples':
+            # Add an organ aggregate to the Elasticsearch request
+            es_search.aggs.bucket('organTypes',
+                                  'terms',
+                                  field='contents.samples.effective_organ.keyword',
+                                  size=config.terms_aggregation_size)
+        else:
+            assert entity_type == 'projects', entity_type
+
+        threshold = config.precision_threshold
+        for agg_name, cardinality in cardinality_aggregations.items():
+            es_search.aggs.metric(agg_name,
+                                  'cardinality',
+                                  field=cardinality + '.keyword',
+                                  precision_threshold=str(threshold))
 
         self._annotate_aggs_for_translation(es_search)
         es_search = es_search.extra(size=0)
         es_response = es_search.execute(ignore_cache=True)
         assert len(es_response.hits) == 0
         self._translate_response_aggs(catalog, es_response)
-        final_response = SummaryResponse(es_response.to_dict())
+
         if config.debug == 2 and logger.isEnabledFor(logging.DEBUG):
             logger.debug('Elasticsearch request: %s', json.dumps(es_search.to_dict(), indent=4))
-        return final_response.return_response().to_json()
+
+        result = es_response.aggs.to_dict()
+        for agg_name in cardinality_aggregations:
+            agg_value = result[agg_name]['value']
+            assert agg_value <= threshold / 2, (agg_name, agg_value, threshold)
```

Comment:
The `terms` aggregation is similarly approximate. We should add a FIXME, (not here but in _create_aggregate) to apply an equivalent sanity check to curb or eliminate approximation errors for the terms aggregation.

https://www.elastic.co/guide/en/elasticsearch/reference/6.8/search-aggregations-bucket-terms-aggregation.html#search-aggregations-bucket-terms-aggregation-approximate-counts